ddtrace/ext/ci_visibility/api.py

```diff
@@ -0,0 +1,435 @@
+        force_finish_children: bool = False,
+    ):
+        log.debug(
+            "Finishing module %s, override_status: %s, force_finish_suites: %s",
+            item_id,
+            override_status,
+            force_finish_children,
+        )
+        core.dispatch(
+            "ci_visibility.module.finish", (CIModule.FinishArgs(item_id, override_status, force_finish_children),)
+        )
+
+
+class CISuite(_CIVisibilityAPIBase):
+    class DiscoverArgs(NamedTuple):
+        suite_id: CISuiteId
+        codeowners: Optional[List[str]] = None
+        source_file_info: Optional[CISourceFileInfo] = None
+
+    class FinishArgs(NamedTuple):
+        suite_id: CISuiteId
+        force_finish_children: bool = False
+        override_status: Optional[CITestStatus] = None
+        is_itr_skipped: bool = False
+
+    @staticmethod
+    @_catch_and_log_exceptions
+    def discover(
+        item_id: CISuiteId,
+        codeowners: Optional[List[str]] = None,
+        source_file_info: Optional[CISourceFileInfo] = None,
+    ):
+        """Registers a test suite with the CI Visibility service."""
+        log.debug("Registering suite %s, source: %s", item_id, source_file_info)
+        core.dispatch("ci_visibility.suite.discover", (CISuite.DiscoverArgs(item_id, codeowners, source_file_info),))
+
+    @staticmethod
+    @_catch_and_log_exceptions
+    def start(item_id: CISuiteId):
+        log.debug("Starting suite %s", item_id)
+        core.dispatch("ci_visibility.suite.start", (item_id,))
+
+    @staticmethod
+    @_catch_and_log_exceptions
+    def finish(
+        item_id: CISuiteId,
+        force_finish_children: bool = False,
+        override_status: Optional[CITestStatus] = None,
+        is_itr_skipped: bool = False,
+    ):
+        log.debug(
+            "Finishing suite %s, override_status: %s, force_finish_children: %s",
+            item_id,
+            force_finish_children,
+            override_status,
+        )
+        core.dispatch(
+            "ci_visibility.suite.finish",
+            (CISuite.FinishArgs(item_id, force_finish_children, override_status, is_itr_skipped),),
+        )
+
+    @staticmethod
+    @_catch_and_log_exceptions
+    def mark_itr_skipped(item_id: CITestId, force_finish_children: bool = False):
+        log.debug("Marking test %s as skipped by ITR", item_id)
+        CISuite.finish(item_id, force_finish_children, CITestStatus.SKIP, is_itr_skipped=True)
+
+    @staticmethod
+    @_catch_and_log_exceptions
+    def add_coverage_data(item_id: CITestId, coverage_data: Dict[str, Any]):
+        log.debug("Adding coverage data for suite %s: %s", item_id, coverage_data)
+
+
+class CITest(_CIVisibilityAPIBase):
+    class DiscoverArgs(NamedTuple):
+        test_id: CITestId
+        codeowners: Optional[List[str]] = None
+        source_file_info: Optional[CISourceFileInfo] = None
+
+    @staticmethod
+    @_catch_and_log_exceptions
+    def discover(
```

Comment:
just out of curiosity: at what stage does attachment of test to test suite happens?

Reply:
Part of the stateless aspect is that there is actually no "attachment" from the consumer side of the API: the user of the API has to create consistent ID objects. While that does arguably open the door for bugs (eg: if the ID objects aren't consistently generated), I think it's pretty easy to get right.

"Inside" CI Visibility, the attachment is done at discovery time. I do plan on adding a "non-discovery" mode where the attachment will happen at test start time, but even in those cases, the correctness of the IDs passed in is what determines said attachment.

I preferred that over the idea of having the user have to stash objects somewhere (eg: as attributes on the testrunner's objects).

Reply:
interesting approach, thank you for explanation!